File: packages/gg-sandbox-csharp/src/Gg/Objects/programs.cs
using System.Collections.Generic;

namespace Gg;

/// <summary>
/// fetch a program you already ran, and hand a patched copy back to be run
/// </summary>
/// <remarks>
/// Under responses-as-code a whole turn is a program, so a trivial mistake in a long one costs the
/// whole program again. gg keeps the source of every program it ran for you: fetch one, patch it with
/// ordinary string work, and hand it back.
/// </remarks>
public static class programs
{
    /// <inheritdoc cref="Internal.ObjectDirectory.List"/>
    public static IReadOnlyList<FunctionSummary> List() => Internal.ObjectDirectory.List("programs");

    /// <summary>The programs you have run, oldest first.</summary>
    /// <remarks>
    /// Each entry carries the shape of a program rather than its source — a directory that inlined
    /// sixty lines per entry would put the whole session back into the one place that exists to avoid
    /// re-reading it. Empty before your first program has been recorded.
    /// </remarks>
    /// <returns>one summary per program you have run, oldest first.</returns>
    /// <exception cref="ToolException">
    /// <see cref="ToolErrorCode.Unavailable"/> when this agent keeps no program library.
    /// </exception>
    public static IReadOnlyList<ProgramSummary> History()
    {
        Internal.Wire.Check(Internal.Native.History(
            out var turns,
            out var lines,
            out var chars,
            out var ok,
            out var errors));
        var summaries = new ProgramSummary[turns.Length];
        for (var index = 0; index < turns.Length; index++)
        {
            summaries[index] = new ProgramSummary(
                turns[index],
                lines[index],
                chars[index],
                ok[index],
                errors[index]);
        }
        return summaries;
    }

    /// <summary>The exact source of one program, as it was run.</summary>
    /// <remarks>
    /// "As it was run" is the contract that matters: when a turn's program was itself handed over by an
    /// earlier one, what is kept is the program that executed rather than the few lines that asked for
    /// it. So fetching, patching and re-running composes.
    /// </remarks>
    /// <param name="turn">
    /// The turn to fetch, as <see cref="ProgramSummary.Turn"/> reports it. Leave it out for the most
    /// recent one.
    /// </param>
    /// <returns>that program's source.</returns>
    /// <exception cref="ToolException">
    /// <see cref="ToolErrorCode.NotFound"/> — naming the turns that are held — for a turn that ran no
    /// program, or one the library's retention has already dropped.
    /// </exception>
    public static string Get(uint? turn = null)
    {
        Internal.Wire.Check(Internal.Native.GetProgram(Internal.Wire.Slot(turn), out var source));
        return source;
    }

    /// <summary>Hand gg a program to run in place of this one.</summary>
    /// <remarks>
    /// <para>
    /// It is <b>registered, not performed</b>. The call returns and the rest of your program still
    /// runs; only when it has ended does gg compile and run what you handed over. Everything the
    /// registering program did stands — its calls, its views — and the program that runs next sees
    /// exactly the world it left behind.
    /// </para>
    /// <code>
    /// var previous = programs.Get();
    /// programs.Rerun(previous.Replace("--release", "--debug"));
    /// </code>
    /// <para>
    /// The first call stands and a second is refused: a silently replaced program is a change you
    /// cannot see. The request is revoked if your program then fails, on the same rule an ending is.
    /// </para>
    /// </remarks>
    /// <param name="source">The program to run instead. Blank is refused.</param>
    /// <exception cref="ToolException">
    /// <see cref="ToolErrorCode.InvalidArgument"/> for a blank source, and
    /// <see cref="ToolErrorCode.Refused"/> for a second call in one turn.
    /// </exception>
    public static void Rerun(string source) => Internal.Wire.Check(Internal.Native.Rerun(source));
}

File: packages/gg-sandbox-java/src/gg/programs/Programs.java
package gg.programs;

import gg.ToolError;
import gg.ToolErrorCode;
import gg.internal.Coding;
import gg.internal.Read;
import gg.internal.Value;
import java.util.List;
import java.util.Optional;

/**
 * Fetch a program this session already ran, and hand a patched copy back to be run.
 *
 * <p>Under responses as code a reply is a whole program, so a one-character mistake in a sixty-line
 * program costs the sixty lines again. The library makes the fix proportional to the mistake: fetch
 * what ran, patch it with ordinary string work, hand it back.
 *
 * <pre>{@code
 * Programs.rerun(Programs.get().replace("Files.readfile", "Files.readFile"));
 * }</pre>
 *
 * @ggmodule programs
 */
public final class Programs {
    private Programs() {
    }

    /**
     * Every program this session has run, oldest first.
     *
     * <p>It lists shapes rather than sources, so a directory of forty programs costs a few lines
     * each. The list survives a compaction, which makes it the way back to a program whose text has
     * left the context window. A session that has run nothing yet gets an empty list rather than a
     * failure.
     *
     * @return every program this session has run, oldest first
     * @ggop programs.history
     */
    public static List<ProgramSummary> history() {
        return Read.programSummaries(Coding.call("programs.history"));
    }

    /**
     * The exact source of one program this session ran, as a string; with no turn, the most recent.
     *
     * <p>The first half of fixing a program without rewriting it: get what ran, patch it with
     * ordinary string work, hand the result to {@link #rerun}. What comes back is the program that
     * executed — so where a turn's program was itself handed over by {@link #rerun}, this is the
     * program that ran rather than the few lines that asked for it, and fetch-patch-run composes
     * turn after turn.
     *
     * @return the source of the last program this session ran
     * @throws ToolError {@link ToolErrorCode#NOT_FOUND}, naming the turns that are held, when
     *     nothing has run yet.
     * @ggop programs.get
     */
    public static String get() {
        return Coding.call("programs.get", Value.none()).text();
    }

    /**
     * The program that ran on one particular turn, rather than the most recent one.
     *
     * @param turn The turn whose program to fetch, as {@link #history} reports it.
     * @return the source of the program that ran on that turn
     * @throws ToolError {@link ToolErrorCode#NOT_FOUND}, naming the turns that are held, for a turn
     *     that ran no program or one old enough that the library has dropped it.
     * @ggop programs.get
     */
    public static String get(int turn) {
        return Coding.call("programs.get", Value.of(turn)).text();
    }

    /**
     * Hand gg a program to run in place of this one.
     *
     * <p>This program finishes, then gg compiles and runs {@code source} as the turn's program.
     * Nothing is undone: every call already made stands, and the program that runs next sees the
     * world this one left behind — so a hand-over belongs before work that should not happen twice.
     *
     * <p>The first call stands, because a silently replaced program is a change nobody can see. A
     * program that then fails cancels its hand-over along with everything else it decided, and the
     * turn is an ordinary error turn instead. Chains are bounded: one hand-over per turn, and the
     * program handed over does the work.
     *
     * @param source The program to run in place of this one, as Java statements. It may not be
     *     blank.
     * @throws ToolError {@link ToolErrorCode#REFUSED} for a second hand-over in one turn, and
     *     {@link ToolErrorCode#INVALID_ARGUMENT} for a blank source.
     * @ggop programs.rerun
     */
    public static void rerun(String source) {
        Coding.call("programs.rerun", Value.of(source));
    }

    // -------------------------------------------------------------------------------------------
    // The type the library hands back
    // -------------------------------------------------------------------------------------------

    /**
     * One program this session already ran, as the history lists it.
     *
     * <p>It describes the program's shape, never its source: a directory that inlined every program
     * would put the whole session back in the context window, which is the one thing the library
     * exists to avoid.
     *
     * @param turn The turn it ran on — what {@link Programs#get(int)} takes.
     * @param lines How many lines of source it was.
     * @param chars How many characters of source it was.
     * @param ok Whether it ran to its end, with no uncaught failure and no ceiling stopping it.
     * @param error The error it ended with, where it did not run to its end.
     */
    public record ProgramSummary(int turn, int lines, int chars, boolean ok,
            Optional<String> error) {

        /**
         * Fetch this program's source, which is {@link Programs#get(int)} on the turn it ran.
         *
         * @return the source of the program that ran on this turn
         * @throws ToolError {@link ToolErrorCode#NOT_FOUND} when the library has dropped it.
         * @ggalias programs.get
         */
        public String source() {
            return Programs.get(turn);
        }
    }
}
